### PR TITLE
Re-write the doc index page

### PR DESCRIPTION
**Pull request by steveklabnik**
Clarify and reorganize.

Add the section for the bookshelf. More to come here in the near future!

Part of #39588

### CODE REVIEW

**Comment by QuietMisdreavus** [comment]
Nit: If I remember right, making the link text for all those links simple the word "here" doesn't work well with screen readers, because if you're jumping from link to link, all you hear is the word "here" repeatedly.

File: src/doc/index.md [hunk 1]
```diff
@@ -6,32 +6,44 @@ nav {
 }
 </style>
 
-This is an index of the documentation included with the Rust
-compiler. For more comprehensive documentation see [the
-website](https://www.rust-lang.org).
+This page is an overview of the documentation included with your Rust install.
+Other unofficial documentation may exist elsewhere; for example, the [Rust
+Learning] project collects documentation from the community, and [Docs.rs]
```
Comment (GuillaumeGomez) [inline src/doc/index.md]: `docs.rs`?

File: src/doc/index.md [hunk 2]
```diff
@@ -6,32 +6,44 @@ nav {
...
+diagnostics from the compiler on those errors. We also produce those
+diagnostics on the web [here][err], if you prefer to read them that way.
+
+## The Rust Bookshelf
+
+Rust provides a number of book-length sets of documentation, collectively
+nicknamed 'The Rust Bookshelf.'
+
+* [The Rust Programming Language][book] teaches you how to program in Rust.
+* [The Rustinomicon][nomicon] is your guidebook to the dark arts of unsafe Rust.
```
Comment (petrochenkov) [inline src/doc/index.md]: typo Rustinomicon -> Rustonomicon

File: src/doc/index.md [hunk 3]
```diff
@@ -6,32 +6,44 @@ nav {
...
+stabilization; however, we are still back-filling things that landed before
+then. That work is being tracked [here][38643].
 
-[**Compiler Error Index**][err]. Extended explanations of
-the errors produced by the Rust compiler.
+## Extended Error Documentation
 
-[book]: book/index.html
-[ref]: reference.html
+All of Rust's errors code with error codes, and you can request extended
```
Comment (frewsxcv) [inline src/doc/index.md]: Is this supposed to be "come with error codes"?

File: src/doc/index.md (hunk 1 above)
Comment (steveklabnik) [inline src/doc/index.md]: @onur, how should we refer to your project here?

File: src/doc/index.md (hunk 3 above)
Comment (steveklabnik) [inline src/doc/index.md]: yes, thanks!

**Comment by steveklabnik** [comment]
@QuietMisdreavus good call, and it's not great SEO either.

@camlorn , do you have opinions here?

File: src/doc/index.md (hunk 1 above)
Comment (onur) [inline src/doc/index.md]: @steveklabnik I think this referral is enough, but it's up to you.

File: src/doc/index.md (hunk 3 above)
Comment (ollie27) [inline src/doc/index.md]: This isn't true though, there are errors which don't have error codes like `error: main function not found`.

**Comment by camlorn** [comment]
@steveklabnik 
This is a tough one.  For me, it's fine.  I'm not sure how common using tab actually is, but I don't and consequently get them in context.

I wouldn't notice, basically.  But the thing is that you're asking an extreme power user this question, and honestly I have no idea.  It's probably best to change them, but I'd never have thought to open an issue or PR actually changing them after the fact.

take that for what it's worth.  Sorry I can't be more helpful here.

**Comment by steveklabnik** [comment]
> take that for what it's worth. Sorry I can't be more helpful here.

It's all good! Even this response is helpful here.

**Comment by steveklabnik** [comment]
Review concerns should be addressed.

**Comment by frewsxcv** [comment]
Looks great! @bors r+ rollup

**Pull request merged**